File: 2020/day02-1.ts
type Entity = [number, number, string, string];

function solution(content: string) {
  return content
    .split("\n")
    .map((item) => parse(item))
    .filter((item) => isValid(item)).length;
}

function isValid(entity: Entity) {
  const [min, max, char, password] = entity;
  const filtered = password.split("").filter((ch) => ch === char);
  return min <= filtered.length && filtered.length <= max;
}

function parse(line: string): Entity {
  const match = line.match(/(\d+)\-(\d+) ([a-z]): ([a-z]+)/);
  if (match === null) {
    throw Error(`Invalid line: ${line}`);
  }
  return [Number(match[1]), Number(match[2]), match[3], match[4]];
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2020/day02_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any

type Entity = tuple[int, int, str, str]


def solution(content: list[str]) -> Any:
    return len([True for line in content if is_valid(parse(line))])


def is_valid(entity: Entity):
    minimum, maximum, char, password = entity
    return minimum <= len([True for ch in password if ch == char]) <= maximum


def parse(line: str) -> Entity:
    match = re.match(r"(\d+)-(\d+) ([a-z]): ([a-z]+)", line)
    if match is None:
        raise ValueError(f"Invalid line: {line}")
    return (int(match.group(1)), int(match.group(2)), match.group(3), match.group(4))


def main() -> None:
    with open("day02.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
